contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1283,146 +1287,299 @@ sp_describe_undeclared_parameters_internal(PG_FUNCTION_ARGS)
+			"END  "
+			"END "
+		"WHEN \'varbinary\' THEN "
+			"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  = -1 THEN 65535 "
+			"ELSE CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  "
+			"END "
+		"WHEN \'varchar\' THEN "
+			"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  = -1 THEN 65535 "
+			"ELSE CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  "
+			"END "
+		"WHEN \'decimal\' THEN 17 "
+		"WHEN \'numeric\' THEN 17 "
+		"WHEN \'xml\' THEN 8100 "
+		"WHEN \'image\' THEN 2147483647 "
+		"WHEN \'text\' THEN 2147483647 "
+		"WHEN \'ntext\' THEN 2147483646 "
+		"ELSE CAST( CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  AS INT ) "
 		"END "					/* AS "suggested_tds_length" */
-		"FROM sys.objects O, sys.columns C, sys.types T, sys.types T2 "
-		"WHERE O.object_id = C.object_id "
-		"AND C.user_type_id = T.user_type_id "
-		"AND C.name = \'%s\' COLLATE sys.database_default " /* -- INPUT column name */
-		"AND T.system_type_id = T2.user_type_id "	/* -- To get system dt
-													 * name. */
-		"AND O.name = \'%s\' COLLATE sys.database_default " /* -- INPUT table name */
-		"AND O.schema_id = %d " /* -- INPUT schema Oid */
-		"AND O.type = \'U\'";	/* -- User tables only for the time being */
-
-		char	   *query = psprintf(tempq,
-									 undeclaredparams->targetcolnames[undeclaredparams->paramindexes[call_cntr]],
-									 undeclaredparams->tablename,
-									 undeclaredparams->schemaoid);
-
+		"FROM sys.types T2, pg_attribute a, pg_type t "
+		", sys.translate_pg_type_to_tsql(a.atttypid) AS tsql_type_name "
+		", sys.translate_pg_type_to_tsql(t.typbasetype) AS tsql_base_type_name "
+        "WHERE a.attrelid = %d "
+        "AND a.atttypid = T2.user_type_id " 
+        "AND T2.system_type_id = T2.user_type_id "
```

Comment:
By this, Are we restricting the output to system types only?

Reply:
Yes, currently our implementation of this sp is limited to system types only as this sp is not fully implemented yet. User defined types produces unexpected output. Which is not addressed here as this task only focuses on improving the perf. For the correctness of output, we have other tasks.